Support unconfirmed resource for nodes to improve scheduling fairness

wilfred-s opened the pull request:
This change adds the scheduling node as a permanent object to the
scheduler. The node tracks allocating and preempting resources and is
directly linked to a cache node. A scheduling node is created after the
cache node is created.

This is a reimplementation of PR #63, @TaoYang526 please check this as you did that one.

TaoYang526 commented on pkg/scheduler/drf_preemption_policy.go:
```diff
@@ -114,14 +116,14 @@ type singleNodePreemptResult struct {
+    // ignoring anything below 0 as we have more available than requested
+    // don't count at what is already marked for preemption (that is still used)
+    // the scheduling node's available resource takes into account what is being allocated
+    resourceToPreempt := resources.SubEliminateNegative(node.getAvailableResource(), candidate.AllocatedResource)
```
resourceToPreempt should be (requested - available), right?

wilfred-s replied:
Yes you're right fixed that

TaoYang526 commented on pkg/scheduler/scheduling_node.go:
```diff
@@ -27,67 +27,158 @@ import (
 }
 
 func NewSchedulingNode(info *cache.NodeInfo) *SchedulingNode {
+	// save guard against panic
```
Typo: save -> safe

wilfred-s replied:
fixed

yangwwei commented on pkg/scheduler/scheduling_node.go:
```diff
@@ -27,67 +27,158 @@ import (
+    defer sn.lock.Unlock()
+
+    sn.availableDirty = true
+	sn.allocatingResource.AddTo(proposed)
```
NIT, extra space at this line.

wilfred-s replied:
fixed, made sure we use tabs everywhere

yangwwei commented on pkg/scheduler/scheduling_node.go:
```diff
@@ -27,67 +27,158 @@ import (
+
+// Update the number of resource proposed for allocation on this node
+func (sn *SchedulingNode) incAllocatingResource(proposed *resources.Resource) {
+	sn.lock.Lock()
```
NIT. extra space

wilfred-s replied:
fixed removed spaces and replaced with tab

yangwwei commented on pkg/scheduler/scheduling_node.go:
```diff
@@ -27,67 +27,158 @@ import (
+			zap.String("nodeId", sn.NodeId),
+			zap.Any("total unconfirmed", newAllocating))
+		sn.availableDirty = true
+		sn.allocatingResource = newAllocating
```
Looks like this can be simplified to:

```
availableDirty = true
available := sn.getAvailableResource()
resources.FitIn(available, delta)
  sn. incAllocatingResource(delta)
```

wilfred-s replied:
`available` is updated with the preempting resources during the preemption phase so we can not use that directly from the call. The delta is the _increase_ to what we have.
We cannot ignore the currently already in progress but not confirmed yet allocations which are stored `allocatingResource`.  The `newAllocating` has to fit in `available`, we cannot shortcut this.

yangwwei replied:
Fair point.

yangwwei commented on pkg/scheduler/tests/scheduler_smoke_test.go:
```diff
@@ -1364,131 +1456,124 @@ partitions:
     assert.Equal(t, int(totalNode2Resource), 0)
 }
 
-//func TestScheduleBestEffortRequests(t *testing.T) {
```
can we do not remove this for now?
I think we need some tests to cover best effort request, can't remember what was exactly the reason this was commented out, but it would be nice to work on this and get this part into test coverage.

wilfred-s replied:
The test is never going to work as we cannot schedule a request with all resources set to 0. To allow that we need to properly think about it. We did a shortcut for QOS pods for K8s as per #54.

We moved the real zero resource support issue out. I opened a new issue for this #75.  When we implement that we need to add new tests that test the implementation we then add.

yangwwei commented on pkg/scheduler/scheduling_node.go:
```diff
@@ -27,67 +27,158 @@ import (
+	allocatingResource *resources.Resource // resources being allocated
+	preemptingResource *resources.Resource // resources considered for preemption
+	cachedAvailable    *resources.Resource // calculated available resources
+	availableDirty     bool                // is the calculated available resource up to date?
```
Is `cachedAvailableResourceExpires` a better name for this? Or any other alternative? `availableDirty` is not very self-explanatory to me. And also, this line deserves more comment to explain why we have this flag, which will help other to understand the idea.

wilfred-s replied:
changed the name to `cachedAvailableUpToDate`

wilfred-s commented:
I'll put in a general comment around the performance of the changes. I did expect a performance impact even compared to the earlier PR #63. So to show what is happening a number of runs in  different branches:
```
---------- master
--- PASS: TestSchedulerThroughput5KNodes (32.80s)
    scheduler_perf_test.go:120: Total time to add 5000 node 1.173216255s, 4261.788889 per second
    scheduler_perf_test.go:167: Total time to allocate 10000 containers in 31.49267957s, 317.534111 per second

---------- unconfirmed (as per PR)
--- PASS: TestSchedulerThroughput5KNodes (40.54s)
    scheduler_perf_test.go:121: Total time to add 5000 node 1.181468682s, 4232.020769 per second
    scheduler_perf_test.go:168: Total time to allocate 10000 containers in 39.242126399s, 254.828189 per second

---------- master with PR63 re-instated
--- PASS: TestSchedulerThroughput5KNodes (32.94s)
    scheduler_perf_test.go:120: Total time to add 5000 node 887.322709ms, 5634.928476 per second
    scheduler_perf_test.go:167: Total time to allocate 10000 containers in 31.92603442s, 313.223994 per second

---------- unconfirmed (dirty flag removed)
--- PASS: TestSchedulerThroughput5KNodes (134.85s)
    scheduler_perf_test.go:121: Total time to add 5000 node 1.118588662s, 4469.918362 per second
    scheduler_perf_test.go:168: Total time to allocate 10000 containers in 2m13.608561151s, 74.845503 per second

---------- unconfirmed (kept dirty flag removed debug logging)
--- PASS: TestSchedulerThroughput5KNodes (39.18s)
    scheduler_perf_test.go:121: Total time to add 5000 node 1.075377802s, 4649.528743 per second
    scheduler_perf_test.go:168: Total time to allocate 10000 containers in 37.95201977s, 263.490588 per second
```

Adding nodes is not become much slower. The test has new debug logging in the `SchedulerNode` and we're adding new objects via events. The change will have an impact. Seeing less than 1% difference between the runs looks good and is negligible. 
Allocation has become slower by about 25%.  That is a large impact. However the last run shows the removal of just the new debug logging in the `SchedulerNode` which removed about 3 seconds (or 30%) of the slowdown.
The `dirty` flag as you can see has a large performance impact. Without the flag we go back to resources and do a lot of extra calculations during the sorting.

I'll address the rest of the comments separately.

wilfred-s commented on pkg/cache/cluster_info.go:
```diff
@@ -551,6 +564,24 @@ func (m *ClusterInfo) processRejectedApplicationEvent(event *cacheevent.Rejected
+    // pass all releases to the scheduler
+    m.EventHandlers.SchedulerEventHandler.HandleEvent(
+        &schedulerevent.SchedulerNodeEvent{
+            PreemptedNodeResources: nodeRes,
```
fixed the comment, preempted resources however are always released they are not allocated or revoked, adding release is thus doubling up.

wilfred-s commented on pkg/cache/cluster_info.go:
```diff
@@ -356,24 +344,32 @@ func (m *ClusterInfo) processNewSchedulableNodes(request *si.UpdateRequest) {
+        // create the equivalent scheduling node
+        m.EventHandlers.SchedulerEventHandler.HandleEvent(
+            &schedulerevent.SchedulerNodeEvent{
+                AddedNode: nodeInfo,
```
That means that we do not have a new node until we have processed all nodes and then stop scheduling for the time it takes to add all nodes to the cluster. This will not show up during the unit tests but when you add or remove a large number of nodes during normal operations that might lead to longer lock times and no scheduling actions during that time.

The way we do it now the scheduling node *should* be ready when we are done. The time that the cache and the scheduler are out of sync is minimal.
I was even looking at pulling in the existing allocations inside the loop. Cache and scheduler should be independent and could even run from different go routines but they must be in sync as soon as possible.

yangwwei replied:
This is not a very strong point to me, because adding a node to cache is not a time-consuming work, we are not expecting any meaningful delay here. Instead, if we can pack them up for event dispatching, it is more efficient (scheduler channel is quite busy handling allocations already) and also the coding style is more consistent.

wilfred-s replied:
A channel should be able to move millions of objects per second. It is not the channel process or the number of objects that causes the channel to fill up it is the speed at which we pull them off and can process each of them. Doing one node at a time makes processing really fast and keeps lock time as low as possible. Having large numbers of nodes, like the 5,000 we do in the test is exceptional. A new rack of nodes is the largest number (50 nodes or so) I would expect in normal operating circumstances.

The other point to make to keep it at single nodes is: if we are really scheduling we do not want to hold the lock to process all the nodes in one batch. That could cause a large pause in scheduling. This gets a far more even spread and other events can be processed in between.

I would even argue that packing up any type of update and pushing batched updates is the wrong way as it can lead to longer pauses. You want to keep it lightweight, with short lock times using highly granular locks not heavy with long lock time and coarse locking.

yangwwei replied:
Make sense to me, thanks!

wilfred-s commented on pkg/cache/cluster_info.go:
```diff
@@ -391,19 +387,30 @@ func (m *ClusterInfo) processNodeActions(request *si.UpdateRequest) {
-                node.setSchedulable(true)
+                // set the state to schedulable
+                nodeInfo.SetSchedulable(true)
+            case si.UpdateNodeInfo_DECOMISSION:
```
done logged cloudera/yunikorn-k8shim#61

wilfred-s commented on pkg/cache/cluster_info.go:
```diff
@@ -527,12 +534,18 @@ func (m *ClusterInfo) processAllocationProposalEvent(event *cacheevent.Allocatio
     allocInfo, err := partitionInfo.addNewAllocation(proposal)
     if err != nil {
+        log.Logger().Error("failed to add new allocation to partition",
+            zap.String("allocationKey", proposal.AllocationKey),
```
fixed

wilfred-s commented on pkg/scheduler/drf_preemption_policy.go:
```diff
@@ -199,12 +201,20 @@ func crossQueuePreemptionAllocate(preemptionPartitionContext *preemptionPartitio
             break
         }
+        log.Logger().Debug("preemption check on node",
+            zap.Any("node", node))
```
correct that should not have been there, removed

wilfred-s commented on pkg/scheduler/drf_preemption_policy.go:
```diff
@@ -199,12 +201,20 @@ func crossQueuePreemptionAllocate(preemptionPartitionContext *preemptionPartitio
 
     if preemptResult == nil {
+        log.Logger().Debug("preemption result nil",
+            zap.Any("candidate", candidate))
```
It is the only way we can see that we cannot preempt any resources for that request.
Without it it will look like we have skipped a candidate

yangwwei replied:
In this case, we at least need to refine this message a bit, logging result is nil will not be very helpful.

wilfred-s replied:
updated the text to explain no preemption is possible

wilfred-s commented on pkg/scheduler/scheduler.go:
```diff
@@ -315,13 +315,26 @@ func (m *Scheduler) processAllocationUpdateEvent(ev *schedulerevent.SchedulerAll
+    // There should never be more than one accepted allocation in the list
+    // See cluster_info.processAllocationProposalEvent()
+    if len(ev.AcceptedAllocations) > 0 {
+        alloc := ev.AcceptedAllocations[0]
```
We cannot get more than 1, if you look in the function that is referenced in the comment we reject anything more than 1 and the slice that is returned is always of a length == 1

I have added more comments in the cluster_info.processAllocationProposalEvent()

wilfred-s commented on pkg/scheduler/scheduler.go:
```diff
@@ -462,10 +475,30 @@ func (m *Scheduler) processDeletePartitionConfigsEvent(event *schedulerevent.Sch
+    }
+    // preempted resources have now been released update the node
+    if event.PreemptedNodeResources != nil {
+        m.clusterSchedulingContext.releasePreemptedResources(event.PreemptedNodeResources)
```
When we add them to the node we do that synchronously *within* the scheduler code. When we confirm the allocation it comes out of the cache.
Using a scheduler event inside the scheduler is not done anywhere we can but that would be a departure of what we have now.

yangwwei replied:
I see the reason now. in events.go, since this defines public events
```
type SchedulerNodeEvent struct {
    AddedNode              interface{}
    RemovedNode            interface{}
    PreemptedNodeResources []NodeResource
}
```
can we rename `NodeResource` to `NodePreemptedResource` to avoid confusion?

wilfred-s replied:
I had thought about keeping it generic so we could reuse it later but after thinking about it longer: preemption for now seems to be the only case we need it. I renamed the object

wilfred-s commented on pkg/scheduler/scheduling_context.go:
```diff
@@ -201,3 +203,83 @@ func (csc *ClusterSchedulingContext) removeSchedulingPartition(partitionName str
+// This is a lock free call: locks are taken while retrieving the node and when updating the node
+func (csc *ClusterSchedulingContext) updateSchedulingNodeAlloc(alloc *commonevents.AllocationProposal) {
+    // get the partition and node (both have to exist to get here)
+    node := csc.GetSchedulingNode(alloc.NodeId, alloc.PartitionName)
```
added nil check with warning message log

wilfred-s commented:
In my testing I have found a major issue in the preemption code: preemption is not tracked against an application or even a candidate allocation. This means that the pre-emption that is being worked on and one that is in progress for a different allocation can not be distinguished. 
This could in turn lead to the scheduler thinking that the node has space and thus not need to preempt while there is no space and preemption should be performed. The allocation then fails as it would push the node over its allocation limit.

This is not something that I can solve in this PR.
The PR now breaks the `TestBasicPreemption`

yangwwei commented:
For the benchmark test result, 
> Allocation has become slower by about 25%. That is a large impact. However, the last run shows the removal of just the new debug logging in the SchedulerNode which removed about 3 seconds (or 30%) of the slowdown.

what does this mean? "removal of just the new debug logging"
Comparing to the master result `32.80s` vs `39.18s`, are they both running under DEBUG logging level?

yangwwei commented:
@wilfred-s do you have any updates on this one?

wilfred-s commented:
Sorry I was taken up with a lot of internal planning and have only had minimal time to work on this:

For the debug logging remark I made: The tests for speed were all run the log level set to DEBUG. The difference between the runs were that I removed the two new lines of debug logging I added in the SchedulerNode code. The rest stayed the same. It shows the overhead of the logging for those two lines when we try to allocate 10,000 containers. In other words if logging is turned up to DEBUG logging the application spends a lot of time writing out the 10,000 complex lines into the log.

wilfred-s commented:
rebased all changes to head of trunk force pushed the changes

yangwwei commented:
All good now, thanks!

wilfred-s commented:
thank you all for the reviews and help on this one
